data-processing/entities/sentences/extractor.py

```diff
@@ -37,6 +41,86 @@
     "ᓸ",
 ]
 
+# regex patterns for detecting entities from the richer tex
+PATTERN_SECTION = r"\\(?:sub)*section[*]*\{[A-Za-z0-9 \{\}\\_.,:-]*\}"
+PATTERN_LABEL = r"\\label\{[A-Za-z0-9 \\_.,:-]*\}"
+PATTERN_ABSTRACT_BEGIN = r"\\begin\{abstract\}"
+PATTERN_ABSTRACT_END = r"\\end\{abstract\}"
+PATTERN_TABLE_BEGIN = r"\\begin\{table[*]*\}"
+PATTERN_TABLE_END = r"\\end\{table[*]*\}"
+PATTERN_FIGURE_BEGIN = r"\\begin\{figure[*]*\}"
+PATTERN_FIGURE_END = r"\\end\{figure[*]*\}"
+PATTERN_REF = r"\\ref\{[A-Za-z0-9 \\_.,:-]*\}"
+#TODO @dykang Fix to capture citations patterns like \cite[GRUs;][]{cho2004}
+PATTERN_CITE = r"\\cite[A-Za-z0-9 \\_.,:-]*\{[A-Za-z0-9 \\_.,:-]*\}"
+PATTERN_SYMBOL = r"\$[A-Za-z0-9\\ \{\}\(\)\[\]^&*_.,\+:\;-\=#]*\$"
+PATTERN_ANY = r"\\[A-Za-z0-9\\_.,:-]*[\{[A-Za-z0-9 \\_.,:-]*\}]*"
+
+# objects for detecting nesting structures from the extended tex
+NESTING_CHARACTERS_MAPPING = {'{':'}', '(':')', '[':']'}
+
+
+def check_nesting_structure(nesting_chars):
+    stack = []
+    for nchar in nesting_chars:
+        if nchar in ['{','(', '[']:
+            stack.append(nchar)
+        elif nchar in ['}',')',']']:
+            if len(stack) == 0:
+                return False
+            else:
+                if NESTING_CHARACTERS_MAPPING[stack[-1]] != nchar:
+                    return False
+                else:
+                    stack.pop()
+        else:
+            return False
+    return True
+
+
+def check_sentence_or_not(tex, entity_dict):
+    if tex is None:
+        return False
+
+    # check whether text in section or not
+    if 'current_section' not in entity_dict or not entity_dict['current_section']:
+        return False
+    # check whether text is caption in figure/table or not. currently, we remove all text including captions in figure/table
+    # TODO @dykang later, distinguish lines in table and lines in captions
+    if 'current_figure' in entity_dict and entity_dict['current_figure']:
+        return False
+    if 'current_table' in entity_dict and entity_dict['current_table']:
+        return False
+
+    # check nesting structures from tex_sub
+    nesting_characters_in_tex = []
+    for c in tex:
+        if c in list(NESTING_CHARACTERS_MAPPING.keys()) + list(NESTING_CHARACTERS_MAPPING.values()):
+            nesting_characters_in_tex.append(c)
+    if not nesting_characters_in_tex:
+        return True
+    return check_nesting_structure(nesting_characters_in_tex)
+
+
+
+def extract_richer_tex(context_tex, tex):
```

Comment:
not entirely sure I understand what's going on  in this function. Is it just splitting context_tex into lines, and removing all but the line before and after? What's richer about the returned context?

Reply:
Got it, it makes sense why you'd want this! And I'm excited to see the new comment. I'm still not entirely sure though what all is in the richer_tex. Is it the line before and after?